### PR TITLE
Outlier Detection and cube blotting in Calspec3

### PR DESCRIPTION
**Pull request by jemorrison**
Fixed outlier detection to correctly call cube_build with the correct channels (MIRI) or gratings (NIRSPEC)
Fixed blot_cube_build.py to correctly blot the median sky cube back to the detector.

In blot_cube_build.py my first attempt was blot_overlap - this took way too long. So I modified the method and created blot_overlap_quick. I can probably delete blot_overlap but I left it in for now.
I think reviewing blot_overlap_quick is sufficient. 

David look at the weighting function in blot_overlap_quick and make sure that is what you want. Also when finding the overlap between the blotting sky and detector I did not search a radius but a distance in x and distance in y. We can change this if you prefer and change it to a radius between the centers (like in cube_build). I was just trying to find all the blotting values that fall on a pixel.

### CODE REVIEW

File: jwst/cube_build/blot_cube_build.py [hunk 1]
```diff
@@ -161,145 +166,185 @@ def blot_images(self):
...
+        # is quicker than looping over detector x,y and finding overlap with large array for
+        # x_cube, y_cube
+        # t0 = time.time()
+        blot_weight = np.zeros_like(blot_flux)
+        num = x_cube.size
+
+        # print('blot_overlap_quick Number of elements on sky mapped to detector going to loop over',num)
+        # TODO need to move this value (roi_det)  to reference file  or at least in spec ?
+        roi_det = 0.5  # 1/2 a pixel
+        for ipt in range(0, num - 1):
```
Comment (philhodge) [inline jwst/cube_build/blot_cube_build.py]: It looks as if you're skipping the last element in `x_cube` and `y_cube`.  Was that intentional?

File: jwst/cube_build/blot_cube_build.py [hunk 2]
```diff
@@ -161,145 +166,185 @@ def blot_images(self):
...
+            # search xcenter and ycenter seperately. These arrays are smallsh.
+            # xcenter size = naxis1 (naxis1/2 for MIRI)
+            # ycenter size = naxis2
+            xdistance = np.absolute(x_cube[ipt] - xcenter)
```
Comment (philhodge) [inline jwst/cube_build/blot_cube_build.py]: Are `x_cube` and `y_cube` 1-D arrays or 2-D?  It looks as if they were created by concatenating 1-D arrays together to make 2-D arrays, but `num` is the total number of elements in `x_cube`, not just the number in the first axis.  If they're 1-D arrays, then these differences are scalars, so you don't need `np.where`.  If they're 2-D arrays, it looks to me as if the loop index will eventually exceed the length of the first axis length of `x_cube` and `y_cube`.

File: jwst/cube_build/cube_build_step.py [hunk 3]
```diff
@@ -206,7 +206,7 @@ def process(self, input):
         self.pipeline = 3
         if self.output_type == 'multi' and len(self.input_filenames) == 1:
             self.pipeline = 2
-
+        print('in cube build the self.pipeline',self.pipeline)
```
[the remark this answers is not in the document]
Comment (jemorrison) [reply jwst/cube_build/cube_build_step.py]: fixed

File: jwst/cube_build/cube_cloud.py [hunk 4]
```diff
@@ -79,17 +79,17 @@ def match_det2cube_msm(naxis1, naxis2, naxis3,
 
     nplane = naxis1 * naxis2
 
-# now loop over the pixel values for this region and find the spaxels that fall
-# withing the region of interest.
+    # now loop over the pixel values for this region and find the spaxels that fall
+    # withing the region of interest.
```
[the remark this answers is not in the document]
Comment (jemorrison) [reply jwst/cube_build/cube_cloud.py]: fixed

File: jwst/outlier_detection/outlier_detection_ifu.py [hunk 5]
```diff
@@ -100,12 +118,17 @@ def do_detection(self):
         exptype = self.input_models[0].meta.exposure.type
         log.info("Performing IFU outlier_detection for exptype {}".format(
                  exptype))
-        for channel in range(len(self.channels)):
-            ch = self.channels[channel]
+        for band in self.ifu_band:
+            if self.instrument == 'MIRI':
+                cubestep = CubeBuildStep(config_file=cube_build_config,
+                                         channel=band,
+                                         single='true')
```
[the remark this answers is not in the document]
Comment (jemorrison) [reply jwst/outlier_detection/outlier_detection_ifu.py]: yes it is. I changed it. But some how that still worked.

File: jwst/cube_build/blot_cube_build.py (hunk 2 above)
Comment (jemorrison) [reply jwst/cube_build/blot_cube_build.py]: x_cube and y_cube are 1D arrays. I can not map all the data in one transform but I have to loop over the slices and I just keep adding the inverse transforms from the slice to the x_cube and y_cube.  So xcenter is an array of scalars and x_cube[ipt) is a float it is the exact x detector location which has decimals of the ra,dec, lambda point on the sky.

Comment (philhodge) [reply jwst/cube_build/blot_cube_build.py]: OK; thanks for the clarification.

File: jwst/outlier_detection/outlier_detection_ifu.py [hunk 6]
```diff
@@ -67,14 +67,30 @@ def __init__(self, input_models, reffiles=None, **pars):
...
+
+    def _find_ifu_coverage(self):
+        self.channels = []
+        self.gratings = []
+        self.instrument = self.input_models[0].meta.instrument.name.upper()
+        n = len(self.input_models)
+        for i in range(n):
+            if self.instrument == 'MIRI':
+                this_channel = (self.input_models[i].meta.instrument.channel)
+                nc = len(this_channel)
```
[the remark this answers is not in the document]
Comment (jemorrison) [reply jwst/outlier_detection/outlier_detection_ifu.py]: so to be clear what I am doing in these lines is taking the channel '12' or '34' and splitting it up
into '1', '2' or '3', '4'
Is that what extend does ?

[the remark this answers is not in the document]
Comment (philhodge) [reply jwst/outlier_detection/outlier_detection_ifu.py]: No; I misunderstood.  extend concatenates lists.  So what you're doing looks like the best way.

**Comment by jemorrison** [comment]
@stscieisenhamer @hbushouse 
Can we merge this into JWST.  Once this is merged the INS branch (David Law) can test it and after testing it they might have suggestions for how to tweak the weighting in creating the blotted image.s  I ran the NIRSPEC and MIRI tests that fails in DMS testing and they now run and created the blotting images for outlier detection.

**Comment by stscieisenhamer** [comment]
Its been approved. However, it looks like you still need to update the changelog. If you could do that, then feel free to merge.

**Comment by jemorrison** [comment]
closes JP-1117 and JP-1138

**Pull request merged**